Libraries/LibWeb/DOM/Node.h

```diff
@@ -528,53 +528,83 @@ class Node : public EventTarget {
     template<typename Callback>
     TraversalDecision for_each_in_inclusive_subtree(Callback callback) const
     {
-        if (auto decision = callback(static_cast<Node const&>(*this)); decision != TraversalDecision::Continue)
-            return decision;
-        for (auto* child = first_child(); child; child = child->next_sibling()) {
-            if (child->for_each_in_inclusive_subtree(callback) == TraversalDecision::Break)
+        Node const* current = this;
+        while (current != nullptr) {
+            TraversalDecision decision = callback(*current);
+            if (decision == TraversalDecision::Break)
                 return TraversalDecision::Break;
+
+            if (decision != TraversalDecision::SkipChildrenAndContinue && current->first_child() != nullptr) {
+                current = current->first_child();
+                continue;
+            }
+            if (current == this)
+                break;
+
+            if (current->next_sibling() != nullptr) {
+                current = current->next_sibling();
+                continue;
+            }
+
+            while (current != this && current->next_sibling() == nullptr)
+                current = current->parent();
+            if (current == this)
+                break;
+
+            current = current->next_sibling();
```

Comment:
Sweet! In my opinion the new code is large enough that it warrants its own template, so you can reuse it for both the const and non-const versions.